src/main/cpp/src/cast_float_to_string.cu

```diff
@@ -0,0 +1,376 @@
+   *
+   * @return The number of decimal places.
+   */
+  __device__ int dissect_value(double value,
+                               unsigned int& integer,
+                               unsigned long long& decimal,
+                               int& exp10)
+  {
+    // normalize step puts value between lower-limit and upper-limit
+    // by adjusting the exponent up or down
+    exp10 = 0;
+    if (value > upper_limit) {
+      int fx = 256;
+      for (int idx = 8; idx >= 0; --idx) {
+        if (value >= upper10[idx]) {
+          value *= lower10[idx];
+          exp10 += fx;
+        }
+        fx = fx >> 1;
+      }
+    } else if ((value > 0.0) && (value < lower_limit)) {
+      int fx = 256;
+      for (int idx = 8; idx >= 0; --idx) {
+        if (value < blower10[idx]) {
+          value *= upper10[idx];
+          exp10 -= fx;
+        }
+        fx = fx >> 1;
+      }
+    }
+    //
+    // int decimal_places = significant_digits - (exp10? 2 : 1);
+    // unsigned long long max_digits = (exp10? fifteen_digits : sixteen_digits);
+    int decimal_places = significant_digits - 1;
+    unsigned long long max_digits = sixteen_digits;
+    double temp_value = value;
+    while (temp_value < 1.0 && temp_value > 0.0) {
+      max_digits *= 10;
+      temp_value *= 10.0;
+      decimal_places++;
+    }
+    integer                 = (unsigned int)value;
+    for (unsigned int i = integer; i >= 10; i /= 10) {
+      --decimal_places;
+      max_digits /= 10;
+    }
+    double diff = value - (double)integer;
+    double remainder = diff * (double)max_digits;
+    decimal          = (unsigned long long)remainder;
+    remainder -= (double)decimal;
+    decimal += (unsigned long long)(2.0 * remainder); // round up
+    if (decimal >= max_digits) {
+      decimal = 0;
+      ++integer;
+      if (exp10 && (integer >= 10)) {
+        ++exp10;
+        integer = 1;
+      }
+    }
+    //
+    while ((decimal % 10) == 0 && (decimal_places > 0)) {
+      decimal /= 10;
+      --decimal_places;
+    }
+    return decimal_places;
+  }
+
+  /**
+   * @brief Main kernel method for converting float value to char output array.
+   *
+   * Output need not be more than (significant_digits + 7) bytes:
+   * 7 = 1 sign, 1 decimal point, 1 exponent ('e'), 1 exponent-sign, 3 digits for exponent
+   *
+   * @param value Float value to convert.
+   * @param output Memory to write output characters.
+   * @return Number of bytes written.
+   */
+  __device__ int float_to_string(double value, char* output)
+  {
+    // check for valid value
+    if (std::isnan(value)) {
+      memcpy(output, "NaN", 3);
+      return 3;
+    }
+    bool bneg = false;
+    if (signbit(value)) {  // handles -0.0 too
+      value = -value;
+      bneg  = true;
+    }
+    if (std::isinf(value)) {
+      if (bneg)
+        memcpy(output, "-Infinity", 9);
+      else
+        memcpy(output, "Infinity", 8);
```

Comment:
```suggestion
      if (bneg) {
        memcpy(output, "-Infinity", 9);
      } else {
        memcpy(output, "Infinity", 8);
      }
```
It is advisable to use braces all the time due to how easy it is to come back and add bugs.